Automattic/wp-calypso
# Hide app banners on mobile apps

#### Changes proposed in this Pull Request

* Add a new `isMobileApp` utility that identifies if Calypso is being accessed from a mobile app by checking that the user agent contains `wp-iphone` ([used on the iOS app user agent](https://github.com/wordpress-mobile/WordPress-iOS/blob/4ac54edd8c4db2978a0122f97e137c755e627d4a/WordPress/Classes/Utility/WPUserAgent.m#L33-L43)) or `wp-android` ([used on the Android app user agent](https://github.com/wordpress-mobile/WordPress-iOS/blob/4ac54edd8c4db2978a0122f97e137c755e627d4a/WordPress/Classes/Utility/WPUserAgent.m#L33-L43)).
* Hide the app banner on requests coming from the mobile app (we're already in the app, so we don't need to show this banner).

<img width="294" alt="Screen Shot 2019-03-28 at 12 50 23" src="https://user-images.githubusercontent.com/1233880/55128787-2cacba00-5158-11e9-9554-933b23476f69.png">

#### Testing instructions

* Checkout this branch and start Calypso locally: `npm start`.
* Open the reader: `calypso.localhost:3000`.
* Change the screen size to match a mobile viewport.
  * On Chrome you can quickly do this on the browser devtools:
<img width="396" alt="Screen Shot 2019-03-28 at 12 54 46" src="https://user-images.githubusercontent.com/1233880/55128997-d7bd7380-5158-11e9-8bbe-751f6249d977.png">

* Note the app banner appears:
<img width="307" alt="Screen Shot 2019-03-28 at 12 57 25" src="https://user-images.githubusercontent.com/1233880/55129086-2965fe00-5159-11e9-90b7-ceed4c0cf6ed.png">

* Use a fake user agent that contains either `wp-iphone` or `wp-android`:
  * iOS example: `Mozilla/5.0 (iPhone; CPU iPhone OS 12_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Mobile/16B91 wp-iphone/12.1`.
  * Android example: `Mozilla/5.0 (Linux; Android 6.0; Android SDK built for x86_64 Build/MASTER; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/44.0.2403.119 Mobile Safari/537.36 wp-android/4.7`.
  * On Chrome you can do this using the Network conditions on the browser devtools:
<img width="404" alt="Screen Shot 2019-03-28 at 13 01 49" src="https://user-images.githubusercontent.com/1233880/55129228-ab562700-5159-11e9-9a75-d05cfe97b65e.png">
<img width="900" alt="Screen Shot 2019-03-28 at 13 02 45" src="https://user-images.githubusercontent.com/1233880/55129288-df314c80-5159-11e9-8278-f14471c9eb2f.png">

* Reload the page.
* Make sure the app banner doesn't show up now.

Fixes #31160
